clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -257,8 +257,13 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, C
                 }
             }
         }
-        if (!expiredBatches.isEmpty())
+        if (!expiredBatches.isEmpty()) {
             log.trace("Expired {} batches in accumulator", count);
+            for (RecordBatch batch : expiredBatches) {
+                batch.expire();
+                deallocate(batch);
```

Comment:
@ijuma Thank you for the review. You can get a reference to a batch only while holding the deque lock. Since these batches were removed from the deque while holding the deque lock, it should be safe to close and deallocate the batches without a lock. A similar pattern was being used in `abortBatches` (though that should be safe anyway since it is only used during close).